### ISSUE TITLE
TcpStream::shutdown on macOS on localhost yields error

### ISSUE THREAD

**Issue by jonhoo** [tokio-rs/tokio#4665]
**Version**
```
tokio 1.18.1
tokio-macros 1.7.0
```

**Platform**
macOS GitHub Runner 11.6.5

**Description**

Calling `TcpStream::shutdown` on a localhost TCP socket on macOS yields

```
Os { code: 57, kind: NotConnected, message: "Socket is not connected" }
```

even though writes to, and reads from, the socket work just fine.

The code in question is here:

https://github.com/jonhoo/async-bincode/blob/b003d7d700625d0f94c46df8a65b29f4c71552c6/src/lib.rs#L150-L159

The `AsyncBincodeStream` mostly just turns the individually sent items into `write_all` calls, and does `read` that result in the `Stream` later. It does nothing "funky" with the socket.

What's interesting is that if I remove the `.unwrap()` from

```rust
c.get_mut().shutdown().await.unwrap();
```

The code runs just fine, meaning the connection _does_ get shut down (otherwise the read loop below would block forever). It's just that the shutdown call also returns an error, even though it doesn't do so on any of the other platforms.

**Comment by Darksonn** [comment]
I'm pretty sure that `TcpStream::shutdown` just calls the standard library function with the `how` argument set to `Write`.

**Comment by jonhoo** [comment]
Yeah, it does seem weird, but that's the behavior I'm seeing. Could be that localhost connections are handled differently on macOS somehow?

**Comment by Darksonn** [comment]
Certainly could be the case.

**Comment by soul2eat** [comment]
Also got this error on Ubuntu 16.04.7 LTS. At first everything works fine, but after a while all connections fail with this error, but all data is transferred.

**Comment by soul2eat** [comment]
It may also be available for this test machine.

**Comment by dd-dreams** [comment]
I can't reproduce it. This is the code I tried.
```rust
use tokio::net;
use tokio::io::{AsyncWriteExt, AsyncReadExt};

#[tokio::main]
async fn main() {
    let listener = net::TcpListener::bind("127.0.0.1:0").await.unwrap();
    let addr = listener.local_addr().unwrap();

    let mut connector = net::TcpStream::connect(addr).await.unwrap();

    let (mut receiver, _) = listener.accept().await.unwrap();

    connector.write_u8(1).await.unwrap();
    receiver.read_u8().await.unwrap();

    connector.shutdown().await.unwrap();
}
```

I also tried the person's crate tests, also all pass.

**Comment by Darksonn** [comment]
I'm going to close this as "needs more info".

**Comment by soundofspace** [comment]
Coming back to this with some more info: this is a Mac OS and BSD specific problem where if shutdown is called twice on a socket it will throw `ENOTCONN`. [Important detail](https://review.lttng.org/c/lttng-tools/+/4374/9) here is that it will still run all the shutdown events and weakup threads waiting on this socket. 

Fixing this is pretty trivial if you can control all shutdown events, but becomes almost impossible for something like a TcpStream which can be closed by a TCP reset. This bug/issue doesn't happen on Linux because Linux silently ignores this type of error in kernel space and just does a shutdown if it is still needed.

Depending on the exact order of events it can happen quite frequent that this error triggers resulting in a lot of side effects in downstream code/projects. Because of that I believe that this should be handled inside Tokio with something like this for a tcp stream (other streams might also need something similar but so far have only seen this issue with tcp):

```rust
    pub(super) fn shutdown_std(&self, how: Shutdown) -> io::Result<()> {
        let result = self.io.shutdown(how);
        match result {
            Err(err) if err.kind() == ErrorKind::NotConnected => Ok(()),
            x => x
        }
    }
```

Some other rust projects having issues with this behaviour:
- https://github.com/plabayo/rama/issues/504
- https://github.com/vectordotdev/vector/issues/22617

Some other interesting links in general (all the same problem but in different programming languages):
- https://github.com/mono/mono/issues/7368
- https://github.com/kmonsoor/py-amqplib/issues/45
- https://github.com/irods/python-irodsclient/issues/657
- https://github.com/python/cpython/issues/48647

I'm happy to open a PR for this with all of this more worked out, but wanted to get some opinions on this first.

**Comment by Darksonn** [comment]
Thanks a lot for diagnosing this! It sounds reasonable.

**Comment by soundofspace** [comment]
I did quite a lot of testing on this, and this is what is saw:

Closing socket after TCP reset was received and handled by OS:
- Linux: Error code 107 -> `NotConnected`
- Mac: Error code 75 -> `NotConnected`

Closing socket after a TCP reset is actually what lead me here, but seems to be the same "problem" on Mac and Linux. Original bug in Hyper I noticed seems to trigger this reset on Mac and not on Linux (or I just haven't been able to reproduce it there).

What is different on Mac and Linux is how closing a socket works. On Linux it is possible to keep calling `AsyncWriteExt::shutdown(&mut stream)`, only when both sides are closed and the actual socket is closed will it start throwing `NotConnected`. On Mac this can only be called once and will then instantly start throwing.

Seeing how Tokio hides this shutdown behind a common trait and we loose the context that we are using a tcp stream, I still believe this specific error should be handled inside of Tokio. Either in `AsyncWrite::poll_shutdown` for tcp stream, or in `TcpStream::shutdown_std`.

To handle this I see two options:
1. Completely ignore `NotConnected` errors during shutdown for tcp stream
2. Ignore `NotConnected` error only the first time we call shutdown. This handles the problem of a TCP reset happening behinds the scenes of which we have no control, while still keep the "native" behaviour when you call it multiple times.

I'm honestly not sure which one is better, option two seems like the smarter one, but option one is really simple and clean, especially since this error should have little or no impact on the result of a shutdown since we don't do any cleanup here like flushing data.

**Comment by Darksonn** [comment]
I think option 1 is much preferable because it doesn't require storing any additional information anywhere.

**Comment by soundofspace** [comment]
Yea agree with that, will update PR to implement that
